core/shared/src/main/scala/zio/internal/WeakConcurrentBag.scala

```diff
@@ -18,36 +22,44 @@ import scala.annotation.tailrec
  * possible performance of the `add` method, which has to do maintenance of the
  * nursery and occassional garbage collection.
  */
-private[zio] class WeakConcurrentBag[A](nurserySize: Int, isAlive: A => Boolean) {
-  val nursery   = RingBuffer[A](nurserySize)
-  val graduates = Platform.newConcurrentSet[WeakReference[A]]()(zio.Unsafe.unsafe)
+private[zio] class WeakConcurrentBag[A <: AnyRef](nurserySize: Int, isAlive: IsAlive[A]) {
+  private[this] val nursery   = MutableConcurrentQueue.boundedPartitioned[WeakReference[A]](nurserySize)
+  private[this] val graduates = Platform.newConcurrentSet[WeakReference[A]](nursery.capacity * 2)(Unsafe.unsafe)
+  private[this] val gcStatus  = new AtomicBoolean(false)
 
   /**
    * Adds a new value to the weak concurrent bag, graduating nursery occupants
    * if necessary to make room.
    */
-  final def add(a: A): Unit =
-    if (!nursery.offer(a)) {
-      graduate()
-      add(a)
+  final def add(a: A): Unit = {
+    val ref     = new WeakReference(a)
```

Comment:
I can't argue with the performance improvement. Although two things:

1. With this change, does the concept of a nursery even make sense and buy us anything, or should we just be using a concurrent set to store everything?
2. It's possible that a `ReferenceQueue` could help us cut down on the number of allocations of `WeakReference` by reusing old ones.

Reply:
I tried both cases, and they both performed significantly worse than the current implementation. The current implementation relies heavily that we _won't_ need to move things into long-term storage - but that the objects will no longer be alive by the time we need to clean up the RingBuffer queues.

Also, I removed the periodic GC from ZScheduler.Supervisor, into a dedicated thread that's spawned by the `WeakConcurrentBag` itself. I think that makes sense now that we won't have the supervisor on by default. I set GC interval to 5 seconds for no real reason other than I think that it's an interval that makes sense (?). I also set a lower thread priority (4) to the GC thread, just to ensure that it doesn't interfere with throughput during times of high CPU utilization